packages/components/src/theme/ThemeUtils.ts

```diff
@@ -119,6 +142,162 @@ export function getThemePreloadData(): ThemePreloadData | null {
+  let parenLevel = 0;
+
+  while (start > -1) {
+    parenLevel = 1;
+    let i = start + cssVarPrefix.length;
+    for (; i < value.length; i += 1) {
+      if (value[i] === '(') {
+        parenLevel += 1;
+      } else if (value[i] === ')') {
+        parenLevel -= 1;
+      }
+
+      if (parenLevel === 0) {
+        ranges.push([start, i]);
+        break;
+      }
+    }
+
+    if (parenLevel !== 0) {
+      log.error('Unbalanced parentheses in css var expression', value);
+      return [];
+    }
+
+    start = value.indexOf(cssVarPrefix, i + 1);
+  }
+
+  return ranges;
+}
+
+/**
+ * Make a copy of the given object replacing any css variable expressions
+ * contained in its prop values with values resolved from the given HTML element.
+ * Variables that resolve to color strings will also be normalized to rgb or
+ * rgba color strings.
+ *
+ * Note that the browser will force a reflow when calling `getComputedStyle` if
+ * css properties have changed. In order to avoid a reflow for every property
+ * check we use distinct setup, resolve / normalize, and cleanup passes:
+ * 1. Setup - Create a tmp element and set all css props we want to evaluate
+ * 2. Resolve / Normalize - Evaluate all css props via `getPropertyValue` calls
+ *    and replace the original expressions with resolved values. Also normalize
+ *    css colors to rgb/a.
+ * 3. Cleanup - Remove the tmp element
+ * @param record An object whose values may contain css var expressions
+ * @param targetElement The element to resolve css variables against. Defaults
+ * to document.body
+ */
+export function resolveCssVariablesInRecord<T extends Record<string, string>>(
+  record: T,
+  targetElement: HTMLElement = document.body
+): T {
+  const perfStart = performance.now();
+
+  // Add a temporary div to attach temp css variables to
+  const tmpPropEl = document.createElement('div');
+  targetElement.appendChild(tmpPropEl);
+
+  const varExpressions = [...extractDistinctCssVariableExpressions(record)];
+
+  // Set temporary css variables for resolving var expressions
+  varExpressions.forEach((varExpression, i) => {
+    const tmpPropKey = `--${TMP_CSS_PROP_PREFIX}-${i}`;
+    tmpPropEl.style.setProperty(tmpPropKey, varExpression);
+  });
+
+  const result = {} as T;
+
+  const computedStyle = window.getComputedStyle(tmpPropEl);
+
+  const resolver = (varExpression: string): string => {
+    const tmpPropKey = `--${TMP_CSS_PROP_PREFIX}-${varExpressions.indexOf(
+      varExpression
+    )}`;
+
+    const resolved = computedStyle.getPropertyValue(tmpPropKey);
+
+    return ColorUtils.normalizeCssColor(resolved);
+  };
+
+  // Resolve the temporary css variables
+  Object.entries(record).forEach(([key, value]) => {
+    result[key as keyof T] = resolveCssVariablesInString(
+      resolver,
+      value
+    ) as T[keyof T];
+  });
+
+  // Remove the temporary css variables
+  tmpPropEl.remove();
+
+  log.debug('Resolved css variables', performance.now() - perfStart, 'ms');
```

Comment:
Side note - perhaps we should add a `.time` method to our Logger, that wraps `console.time`: https://developer.mozilla.org/en-US/docs/Web/API/console/time
I would put that at a finer grain of detail than `debug2` even. Just noting it.